tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        mock_run_group_by.return_value = "group-out"
+        self.instance._group_by = MockProperties.PROP_1
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_run_group_by.assert_called_once_with(
+            mock_obj_list, self.instance._group_mappings
+        )
+
+    @patch("openstack_query.query_parser.QueryParser._build_unique_val_groups")
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_no_group_mappings(
+        self, mock_run_group_by, mock_build_unique_val_groups
+    ):
+        """
+        Tests that run_parser functions expectedly - when giving group_by with no group_mappings
+        Should call build_unique_val_groups first to set group_mappings and then call run_group_by and return
+        """
+        self.instance._group_by = MockProperties.PROP_1
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+        mock_build_unique_val_groups.return_value = "group-mapping-out"
+        mock_run_group_by.return_value = "group-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_build_unique_val_groups.assert_called_once_with(
+            mock_obj_list, group_by_prop=MockProperties.PROP_1
+        )
+        mock_run_group_by.assert_called_once_with(mock_obj_list, "group-mapping-out")
+
+    @patch("openstack_query.query_parser.QueryParser._run_sort")
+    @patch("openstack_query.query_parser.QueryParser._run_group_by")
+    def test_run_parser_with_sort_and_group(self, mock_run_group_by, mock_run_sort):
+        """
+        Tests that run_parser functions expectedly - when giving both group_by and sort_by
+        Should call run_sort method and then run_group_by on that output, then return
+        """
+        self.instance._sort_by = (MockProperties.PROP_1, False)
+        self.instance._group_by = MockProperties.PROP_1
+        self.instance._group_mappings = {"group1": "some-mapping_func"}
+        mock_obj_list = ["obj1", "obj2", "obj3"]
+
+        mock_run_group_by.return_value = "group-out"
+        mock_run_sort.return_value = "sort-out"
+
+        res = self.instance.run_parser(mock_obj_list)
+        self.assertEqual(res, "group-out")
+        mock_run_sort.assert_called_once_with(mock_obj_list, self.instance._sort_by)
+        mock_run_group_by.assert_called_once_with(
+            "sort-out", self.instance._group_mappings
+        )
+
+    @parameterized.expand(
+        [
+            (
+                "string key ascending",
+                {"arg1": False},
+                [
+                    {"arg1": "a", "arg2": 2},
+                    {"arg1": "b", "arg2": 1},
+                    {"arg1": "c", "arg2": 4},
+                    {"arg1": "d", "arg2": 3},
```

Comment:
Rather than pre-set these I would just do a sort in all of these tests too, something like

```
# Note I would change this to just a string or list of strings, so we can just do mock_sort[0]
prop_name = mock_sort_by_specs.keys()[0]
expected_results = sort(mock_input, key = lambda x: x.prop_name
```

Since it's a trivial bug if we get the sort order wrong to find and fix, and it removes a shed-load of boilerplate